integration-tests/src/tests/client/process_blocks.rs

```diff
@@ -5136,3 +5113,183 @@ mod new_contract_loading_cost {
+    use near_primitives::version::ProtocolFeature;
+
+    /// Snapshot-check outcome of action, with given contract deployed first.
+    fn check_outcome_with_and_without_feature(
+        snap_name: &str,
+        action: Action,
+        contract: Vec<u8>,
+        feature: ProtocolFeature,
+    ) {
+        let account: AccountId = "test0".parse().unwrap();
+        let new_protocol_version = feature.protocol_version();
+        let old_protocol_version = new_protocol_version - 1;
+        let epoch_length: BlockHeight = 5;
+        let mut env =
+            test_env_with_contract(epoch_length, old_protocol_version, account.clone(), contract);
+        let signer = InMemorySigner::from_seed(account.clone(), KeyType::ED25519, account.as_str());
+
+        let tx = Transaction {
+            signer_id: signer.account_id.clone(),
+            receiver_id: signer.account_id.clone(),
+            public_key: signer.public_key(),
+            actions: vec![action],
+            nonce: 0,
+            block_hash: CryptoHash::default(),
+        };
+
+        let mut old_result = env.execute_tx(tx.clone(), &signer, epoch_length);
+        let old_outcome = &mut old_result.receipts_outcome[0].outcome;
+        // Overwrite receipt ids, those will change for many different reasons that are not relevant.
+        for receipt in &mut old_outcome.receipt_ids {
+            *receipt = CryptoHash::default();
+        }
+        insta::assert_debug_snapshot!(format!("{snap_name} before {feature:?}"), old_outcome);
+
+        env.upgrade_protocol(new_protocol_version);
+
+        let mut new_result = env.execute_tx(tx, &signer, epoch_length);
+        let new_outcome = &mut new_result.receipts_outcome[0].outcome;
+        for receipt in &mut new_outcome.receipt_ids {
+            *receipt = CryptoHash::default();
+        }
+        insta::assert_debug_snapshot!(format!("{snap_name} with {feature:?}"), new_outcome);
+    }
+
+    /// Test that current_gas_price host function returns the correct value.
+    #[test]
+    fn test_current_gas_price() {
+        let wat_code = &assert_gas_price_contract("current_gas_price", 10u64.pow(9));
+        let contract = wat::parse_str(wat_code).unwrap();
+        check_outcome_with_and_without_feature(
+            "current_gas_price",
+            Action::FunctionCall(FunctionCallAction {
+                method_name: "main".to_string(),
+                args: vec![],
+                gas: 3 * 10u64.pow(14),
+                deposit: 0,
+            }),
+            contract,
+            ProtocolFeature::GasPriceHostFn,
+        );
+    }
+
+    /// Test that purchased_gas_price host function returns the correct value.
+    #[test]
+    fn test_purchased_gas_price_300_tgas() {
+        // With 300 Tgas, pessimistic inflation is 1.03 ^ 61 = 6.06835119717
+        // and with some rounding it gets to 6_068_351_198.
+        let gas = 3 * 10u64.pow(14);
+        let expected_purchase_price = 6_068_351_198;
```

Comment:
I am wondering if we should more explicitly set various configs that affect prices in tests? 

That's tradeoff between:

* a test explicitly setting its inputs to avoid unrelated changes
* a test checking not only the code, but also the actual configuration we use

Reply:
I think it makes sense to specify the gas related parameters, ideally these tests shouldn't fail when we update those values. So I did this:
```
genesis.config.protocol_version = old_protocol_version;
genesis.config.min_gas_price = 10u128.pow(9);
```